### PR TITLE
feat(subscriptions): enable passing `WebSocket.Server` instead of `HttpServer`.

### PR DESCRIPTION
**Pull request by jedwards1211**
For people who need to have multiple WebSocket servers sharing a single HTTP(S) server, the `ws` docs recommend [creating multiple instances of `new WebSocket.Server({ noServer: true })`](https://github.com/websockets/ws#multiple-servers-sharing-a-single-https-server).  This PR makes `installSubscriptionHandlers` accept a `WebSocket.Server` instance (in addition to accepting an http server) to support the multiple WebSocket servers use case.

Real-life example from my code:
```js
      const graphQLWebSocket = new WebSocket.Server({
        noServer: true,
      })
      apolloServer.installSubscriptionHandlers(graphQLWebSocket)

      const symmetryWebSocket = new WebSocket.Server({
        noServer: true,
      })
      symmetryWebSocket.on('connection', socket => new SymmetryServer({socket}))

      httpServer.on('upgrade', (request: IncomingMessage, socket: net$Socket, head: any) => {
        if (request.url.startsWith(SYMMETRY_BROWSER_PATH)) {
          symmetryWebSocket.handleUpgrade(request, socket, head, (ws: WebSocket) => {
            symmetryWebSocket.emit('connection', ws)
          })
        } else if (request.url.startsWith(GRAPHQL_PATH)) {
          graphQLWebSocket.handleUpgrade(request, socket, head, (ws: WebSocket) => {
            graphQLWebSocket.emit('connection', ws)
          })
        } else {
          socket.destroy()
        }
      })
```

TODO:

* [x] Update CHANGELOG.md with your change (include reference to issue & this PR)
* [x] Make sure all of the significant new logic is covered by tests (NOTE: `installSubscriptionHandlers` is not currently tested anywhere)
* [x] Rebase your changes on master so that they can be merged easily
* [x] Make sure all tests and linter rules pass

### CODE REVIEW

**Comment by jedwards1211** [comment]
@abernix can we get this merged?

**Comment by jedwards1211** [comment]
I fixed the conflict in the changelog, but I hope you can merge this soon so I don't have to keep merging conflicts over and over again

**Comment by LmKupke** [comment]
@abernix this would be great to have

**Comment by jedwards1211** [comment]
@abernix please include this in v3

**Comment by vincentdesmares** [comment]
@abernix please include this in v3

**Comment by vincentdesmares** [comment]
I strongly support this pull request as I find very difficult to patch it on a fork. Sadly we use apollo-server-express, which depends on apollo-server-core. So I would have to fork multiple projects in order to use this patch :(

**Comment by jedwards1211** [comment]
@vincentdesmares [here's my current workaround](https://gist.github.com/jedwards1211/f17ca8116f22bd54a3f3d1227b6e6005)

**Comment by vincentdesmares** [comment]
Thank you so much @jedwards1211 !!!

I did not exactly use your solution as my server is instantiated by a sub-library. But it gave me the idea of monkey-patching the already instantiated Object!

```
apiServer.installSubscriptionHandlers = yourInstallSubscriptionHandlers
jobsServer.installSubscriptionHandlers = yourInstallSubscriptionHandlers
```
And it works perfectly!

That will do it until they release the patch :)

**Comment by jedwards1211** [comment]
You're welcome!  Yeah what I did with `Object.create` is probably unnecessarily complicated.

**Comment by jak-pan** [comment]
Please merge this, +1 needed here.

**Comment by OmgImAlexis** [comment]
@jedwards1211 how is the work around meant to be used? 

<details>
<summary>Currently have something like this, cut down for example sake.</summary>

```js
import stoppable from 'stoppable';
import express from 'express';
import http from 'http';
import { ApolloServer } from 'apollo-server-express';
import { graphql } from './graphql';

const app = express();

const graphApp = new ApolloServer(graphql);
graphApp.applyMiddleware({app});

const httpServer = http.createServer(app);
const stoppableServer = stoppable(httpServer);
graphApp.installSubscriptionHandlers(stoppableServer);
```
</details>

**Comment by jedwards1211** [comment]
@OmgImAlexis basically just monkeypatch the `installSubscriptionHandlers` function, and then see my OP for how to set up multiple websocket handlers with it.

**Comment by jedwards1211** [comment]
@martijnwalraven @abernix will this PR ever get merged?

Will we ever get to stop monkeypatching?

**Comment by OmgImAlexis** [comment]
@jedwards1211 sorry to be a bother but how exactly do I do that?

**Comment by jedwards1211** [comment]
```js
import express from 'express'
import http from 'http'
import WebSocket from 'ws'
import { ApolloServer } from 'apollo-server-express'

const app = express()
const httpServer = http.createServer(app)
const graphQLWebSocket = new WebSocket.Server({ noServer: true })
const otherWebSocket = new WebSocket.Server({ noServer: true })

const GRAPHQL_PATH = '/graphql'
const OTHER_WEB_SOCKET_PATH = '/other'

httpServer.on('upgrade', (request, socket, head) => {
  const { url } = request
  if (url.startsWith(GRAPHQL_PATH)) {
    graphQLWebSocket.handleUpgrade(request, socket, head,ws => {
      graphQLWebSocket.emit('connection', ws)
    })
  } else if (url.startsWith(OTHER_WEB_SOCKET_PATH)) {
    otherWebsocket.handleUpgrade(request, socket, head, ws => {
      otherWebsocket.emit('connection', ws)
    })
  } else {
    socket.destroy()
  }
})

const apolloServer = new ApolloServer({ path: GRAPHQL_PATH })
apolloServer.installSubscriptionHandlers = <the big function from my gist>
apolloServer.applyMiddleware({ app, path: GRAPHQL_PATH })
apolloServer.installSubscriptionHandlers(graphQLWebSocket)

otherWebSocket.on('connection', ...)

httpServer.listen(4000)
```

**Comment by veeramarni** [comment]
This PR is opened for a while and the solution seems to be elegant to have own websocket configuration. @abernix @martijnwalraven  can this PR be reviewed quickly and merge it?

**Comment by kulakowka** [comment]
I use multiple graphql schemes on a single http server. I would like to be able to run several web socket servers. I would like this PR to be merged.

**Comment by Sceat** [comment]
Seems really hard to click on the `merge` button, maybe someone should make a helper script

**Comment by jedwards1211** [comment]
@abernix thanks a bunch!

Even https://graphql.org/learn/ still doesn't even mention subscriptions AFAIK, so subs are kind of the second-class citizen of the whole ecosystem.

Food for thought: in one of my projects I've started a pattern of using matching query and subscription fields implement an "observable" queries.  For it to work I have to do a `cache-only` query, a `no-cache` subscription, and do all the cache updates in `onSubscriptionData`.  The backend sends the initial result set as the first subscription event.

**Comment by okeren-cx** [comment]
I've run into the same problem - needing to run two WS on the same http server, but for some reason I can't get the subscriptions to work:

```
const httpServer = createServer(app);
const timeSyncWebSocketServer = new WebSocket.Server({ noServer: true });
const timesync = new TimeSync(timeSyncWebSocketServer);

const graphQLWebSocketServer = new WebSocket.Server({ noServer: true });
graphqlServer.installSubscriptionHandlers(graphQLWebSocketServer);


httpServer.on('upgrade', (request: IncomingMessage, socket, head: any) => {
  const pathname = url.parse(request.url).pathname;

  if (pathname === '/timesync') {
    timeSyncWebSocketServer.handleUpgrade(request, socket, head, (ws: WebSocket) => {
      timeSyncWebSocketServer.emit('connection', ws);
    });
  } else if (pathname === '/graphql') {
    graphQLWebSocketServer.handleUpgrade(request, socket, head, (ws: WebSocket) => {
      graphQLWebSocketServer.emit('connection', ws);
    });
  } else {
    socket.destroy();
  }
});


timesync.listen();
const tournamentsServer = httpServer.listen(app.get('port'), () => {
```

The timesync feature works great, it enters the `.on('connection...` block properly, but for some reason the graphql subscriptions don't work at all. The block for the `/graphql` pathname is executed, so it seems that maybe the `emit` isn't getting picked up? not sure really, there's no error either.
We're using `apollo-server-express v2.14.4`.

With a single websocket the subscriptions work great:
```
const httpServer = createServer(app);
graphqlServer.installSubscriptionHandlers(httpServer);
const server = httpServer.listen(app.get('port'), () => {
```
but once we add another websocket it doesn't. Does anyone have any idea what might be going on or how to debug this?

**Comment by jedwards1211** [comment]
@okeren-cx you might be having issue #4198, check if there are multiple versions of `ws` in your `node_modules`.

**Comment by okeren-cx** [comment]
@jedwards1211 thanks! it was indeed the problem. I ended up monkey patching in the same way you showed here and it worked.

**Comment by timkock** [comment]
@jedwards1211 and others I had this issue and the issue in #4198.

**TL;DR** until some changes are merged or transport is changed the monkey patch is the way to make it work. The version by @jedwards1211 is in typescript, here is a working javascript equivalent. Thanks to all people that did a lot of the digging before 👍 

https://gist.github.com/timkock/b38ae86cf5634c63dc482c7fc1c66be1
